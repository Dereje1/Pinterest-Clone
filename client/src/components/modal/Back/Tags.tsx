import * as React from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { styled } from '@mui/material/styles';
import Chip from '@mui/material/Chip';
import Paper from '@mui/material/Paper';
import IconButton from '@mui/material/IconButton';
import TagIcon from '@mui/icons-material/Tag';
import updateSearch from '../../../actions/search';
import TagsForm from './TagsForm';
import RESTcall from '../../../crud';
import { PinType, tagType } from '../../../interfaces';

export const ListItem = styled('li')(({ theme }) => ({
  margin: theme.spacing(0.5),
}));

interface TagsArrayProps {
  pinInformation: PinType
  commentFormIsOpen: boolean
  updateTags: (tag: string) => void
  closePin: (_: React.SyntheticEvent, forceClose?: boolean) => void
}

const initialTags: tagType[] | null = [];

function TagsArray({
  pinInformation,
  commentFormIsOpen,
  updateTags,
  closePin,
}: TagsArrayProps) {
  const { tags, owns, _id } = pinInformation;
  const [tagData, setTagData] = React.useState(initialTags);
  const [openTagsForm, setOpenTagsForm] = React.useState(false);
  const [suggestedTags, setSuggestedTags] = React.useState([]);

  const dispatch = useDispatch();
  const history = useHistory();

  const fetchSuggestedTags = async () => {
    try {
      const retreivedTags = await RESTcall({ address: '/api/getTags' });
      setSuggestedTags(retreivedTags);
    } catch (error) {
      setSuggestedTags([]);
    }
  };

  React.useEffect(() => {
    setTagData(tags);
  }, [tags]);

  React.useEffect(() => {
    if (openTagsForm) {
      fetchSuggestedTags();
    }
  }, [openTagsForm]);

  const handleAdd = (val: string) => {
    updateTags(`?pinID=${_id}&tag=${val}`);
    setOpenTagsForm(false);
  };

  const handleDelete = (tagToDelete: tagType) => () => {
    updateTags(`?pinID=${_id}&deleteId=${tagToDelete._id}`);
  };

  const handleTagClick = (e: React.SyntheticEvent, tag: string) => {
    closePin(e);
    dispatch(updateSearch(tag, true));
    history.push('/');
  };

  if (tagData && !commentFormIsOpen && openTagsForm) {
    return (
      <TagsForm
        closeTagsForm={() => setOpenTagsForm(false)}
        addTag={handleAdd}
        suggestedTags={suggestedTags}
        exisitingTags={tagData.map((t) => t.tag.toUpperCase())}
      />
    );
  }

  return (
    <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center' }}>
      <Paper
        sx={{
          display: 'flex',
          justifyContent: 'center',
          flexWrap: 'wrap',
          listStyle: 'none',
          p: 0.5,
          m: 1,
        }}
        component="ul"
        elevation={0}
      >
        {tagData && tagData.map((data) => (
          <ListItem key={data._id}>
            <Chip
              color="warning"
              label={data.tag.toUpperCase()}
              onDelete={owns ? handleDelete(data) : undefined}
              variant="outlined"
              size="small"
              onClick={(e) => handleTagClick(e, data.tag)}
            />
          </ListItem>
        ))}

      </Paper>
      <div style={{ marginRight: 3 }}>
        { owns && tagData && tagData.length < 6 && (
          <IconButton onClick={() => setOpenTagsForm(true)} disabled={commentFormIsOpen}>
            <TagIcon />
          </IconButton>
        )}
      </div>
    </div>

  );
}

export default TagsArray;
