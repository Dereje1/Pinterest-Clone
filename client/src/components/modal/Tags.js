import * as React from 'react';
import PropTypes from 'prop-types';
import { styled } from '@mui/material/styles';
import Chip from '@mui/material/Chip';
import Paper from '@mui/material/Paper';
import IconButton from '@mui/material/IconButton';
import TagIcon from '@mui/icons-material/Tag';
import TagsForm from './TagsForm';

const ListItem = styled('li')(({ theme }) => ({
  margin: theme.spacing(0.5),
}));

function TagsArray({
  pinInformation,
  stylingProps,
  commentFormIsOpen,
  updateTags,
}) {
  const { tags, owns, _id } = pinInformation;
  const [tagData, setTagData] = React.useState([]);
  const [openTagsForm, setOpenTagsForm] = React.useState(false);

  React.useEffect(() => {
    setTagData(tags);
  }, [tags]);

  const handleAdd = (val) => {
    updateTags(`?pinID=${_id}&tag=${val}`);
    setOpenTagsForm(false);
  };

  const handleDelete = (tagToDelete) => () => {
    updateTags(`?pinID=${_id}&deleteId=${tagToDelete._id}`);
  };

  if (!commentFormIsOpen && openTagsForm) {
    return (
      <TagsForm
        closeTagsForm={() => setOpenTagsForm(false)}
        stylingProps={stylingProps}
        addTag={handleAdd}
      />
    );
  }

  return (
    <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center' }}>
      <Paper
        sx={{
          display: 'flex',
          justifyContent: 'center',
          flexWrap: 'wrap',
          listStyle: 'none',
          p: 0.5,
          m: 1,
        }}
        component="ul"
        elevation={3}
      >
        {tagData.map((data) => (
          <ListItem key={data._id}>
            <Chip
              color="info"
              label={data.tag.toUpperCase()}
              onDelete={owns ? handleDelete(data) : undefined}
              variant="outlined"
              size="small"
            />
          </ListItem>
        ))}

      </Paper>
      <div style={{ mr: 3 }}>
        { owns && tagData.length < 6 && (
          <IconButton onClick={() => setOpenTagsForm(true)} disabled={commentFormIsOpen}>
            <TagIcon />
          </IconButton>
        )}
      </div>
    </div>

  );
}

export default TagsArray;

TagsArray.propTypes = {
  stylingProps: PropTypes.objectOf(PropTypes.shape).isRequired,
  commentFormIsOpen: PropTypes.bool.isRequired,
  updateTags: PropTypes.func.isRequired,
  pinInformation: PropTypes.objectOf(PropTypes.shape).isRequired,
};
