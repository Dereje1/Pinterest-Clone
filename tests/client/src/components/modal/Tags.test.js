import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { useHistory } from 'react-router-dom';
import Tags, { ListItem } from '../../../../../client/src/components/modal/Tags';
import { pinsStub } from '../../../pinsStub';
import RESTcall from '../../../../../client/src/crud';

jest.mock('../../../../../client/src/crud');

const mockdispatch = jest.fn();
// Mock router hooks
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useHistory: jest.fn(),
}));

// Mock redux hooks
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'), // use actual for all non-hook parts
  useDispatch: jest.fn(() => mockdispatch),
}));

describe('The tags component', () => {
  let props;
  beforeEach(() => {
    props = {
      pinInformation: { ...pinsStub[0], owns: true },
      commentFormIsOpen: false,
      updateTags: jest.fn(),
      closePin: jest.fn(),
    };
  });

  afterEach(() => {
    props = null;
  });

  test('will render with tags for owner', () => {
    const wrapper = shallow(<Tags {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render with tags for non-owner', () => {
    const updatedProps = {
      ...props,
      pinInformation: pinsStub[0],
    };
    const wrapper = shallow(<Tags {...updatedProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render a list item', () => {
    const wrapper = shallow(<ListItem />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will open and close the form to add tags', async () => {
    const wrapper = shallow(<Tags {...props} />);
    const addTags = wrapper.find('ForwardRef(IconButton)');
    let tagsForm = wrapper.find('TagsForm');
    expect(tagsForm.isEmptyRender()).toBe(true);
    // trigger form show and test tags fetch
    await addTags.props().onClick();
    tagsForm = wrapper.find('TagsForm');
    expect(tagsForm.isEmptyRender()).toBe(false);
    expect(RESTcall).toHaveBeenCalledWith({
      address: '/api/getTags',
    });
    expect(tagsForm.props().suggestedTags).toEqual(['TAG 3', 'TAG 4']);
    // trigger form close
    tagsForm.props().closeTagsForm();
    tagsForm = wrapper.find('TagsForm');
    expect(tagsForm.isEmptyRender()).toBe(true);
  });

  test('will add no suggested tags if rest call is rejected', async () => {
    RESTcall.mockImplementationOnce(() => Promise.reject());
    const wrapper = shallow(<Tags {...props} />);
    const addTags = wrapper.find('ForwardRef(IconButton)');
    // trigger form show and add tag
    await addTags.props().onClick();
    const tagsForm = wrapper.find('TagsForm');
    expect(tagsForm.isEmptyRender()).toBe(false);
    expect(RESTcall).toHaveBeenCalledWith({
      address: '/api/getTags',
    });
    expect(tagsForm.props().suggestedTags).toEqual([]);
  });

  test('will add a tag', () => {
    const wrapper = shallow(<Tags {...props} />);
    const addTags = wrapper.find('ForwardRef(IconButton)');
    // trigger form show and add tag
    addTags.props().onClick();
    const tagsForm = wrapper.find('TagsForm');
    tagsForm.props().addTag('a new tag');
    expect(props.updateTags).toHaveBeenCalledWith('?pinID=1&tag=a new tag');
  });

  test('will delete a tag', () => {
    const wrapper = shallow(<Tags {...props} />);
    const chip = wrapper.find({ label: 'TAG 1' });
    // trigger delete tag
    chip.props().onDelete();
    expect(props.updateTags).toHaveBeenCalledWith('?pinID=1&deleteId=tag_id_1');
  });

  test('will search on tag click', () => {
    const push = jest.fn();
    useHistory.mockImplementation(() => ({ push }));

    const wrapper = shallow(<Tags {...props} />);
    const chip = wrapper.find({ label: 'TAG 1' });
    chip.props().onClick();
    expect(mockdispatch).toHaveBeenCalledWith({
      type: 'UPDATE_SEARCH',
      payload: {
        tagSearch: true,
        term: 'tag 1',
      },
    });
    expect(props.closePin).toHaveBeenCalled();
    expect(push).toHaveBeenCalledWith('/');
  });
});
