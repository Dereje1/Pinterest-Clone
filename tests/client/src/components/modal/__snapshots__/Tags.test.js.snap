// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The tags component will render a list item 1`] = `
<ContextProvider
  value={
    {
      "insert": [Function],
      "inserted": {},
      "key": "css",
      "nonce": undefined,
      "registered": {},
      "sheet": StyleSheet {
        "_insertTag": [Function],
        "before": null,
        "container": undefined,
        "ctr": 0,
        "insertionPoint": undefined,
        "isSpeedy": false,
        "key": "css",
        "nonce": undefined,
        "prepend": undefined,
        "tags": [],
      },
    }
  }
>
  <Insertion
    cache={
      {
        "insert": [Function],
        "inserted": {},
        "key": "css",
        "nonce": undefined,
        "registered": {},
        "sheet": StyleSheet {
          "_insertTag": [Function],
          "before": null,
          "container": undefined,
          "ctr": 0,
          "insertionPoint": undefined,
          "isSpeedy": false,
          "key": "css",
          "nonce": undefined,
          "prepend": undefined,
          "tags": [],
        },
      }
    }
    isStringTag={true}
    serialized={
      {
        "map": undefined,
        "name": "1kxrbkv",
        "next": undefined,
        "styles": "margin:4px;",
        "toString": [Function],
      }
    }
  />
  <li
    className="css-1kxrbkv"
  />
</ContextProvider>
`;

exports[`The tags component will render with tags for non-owner 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "justifyContent": "space-between",
    }
  }
>
  <ForwardRef(Paper)
    component="ul"
    elevation={0}
    sx={
      {
        "display": "flex",
        "flexWrap": "wrap",
        "justifyContent": "center",
        "listStyle": "none",
        "m": 1,
        "p": 0.5,
      }
    }
  >
    <Styled(li)
      key="tag_id_1"
    >
      <ForwardRef(Chip)
        color="warning"
        label="TAG 1"
        onClick={[Function]}
        size="small"
        variant="outlined"
      />
    </Styled(li)>
    <Styled(li)
      key="tag_id_2"
    >
      <ForwardRef(Chip)
        color="warning"
        label="TAG 2"
        onClick={[Function]}
        size="small"
        variant="outlined"
      />
    </Styled(li)>
  </ForwardRef(Paper)>
  <div
    style={
      {
        "marginRight": 3,
      }
    }
  />
</div>
`;

exports[`The tags component will render with tags for owner 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "justifyContent": "space-between",
    }
  }
>
  <ForwardRef(Paper)
    component="ul"
    elevation={0}
    sx={
      {
        "display": "flex",
        "flexWrap": "wrap",
        "justifyContent": "center",
        "listStyle": "none",
        "m": 1,
        "p": 0.5,
      }
    }
  >
    <Styled(li)
      key="tag_id_1"
    >
      <ForwardRef(Chip)
        color="warning"
        label="TAG 1"
        onClick={[Function]}
        onDelete={[Function]}
        size="small"
        variant="outlined"
      />
    </Styled(li)>
    <Styled(li)
      key="tag_id_2"
    >
      <ForwardRef(Chip)
        color="warning"
        label="TAG 2"
        onClick={[Function]}
        onDelete={[Function]}
        size="small"
        variant="outlined"
      />
    </Styled(li)>
  </ForwardRef(Paper)>
  <div
    style={
      {
        "marginRight": 3,
      }
    }
  >
    <ForwardRef(IconButton)
      disabled={false}
      onClick={[Function]}
    >
      <Memo(ForwardRef(TagIcon)) />
    </ForwardRef(IconButton)>
  </div>
</div>
`;
